### ISSUE TITLE
API broken because of defaults?

### ISSUE THREAD

**Issue by jakob-r** [mboecker/lookup-server#36]
1. The naming here is a bit missleading:
https://github.com/mboecker/lookup-server/blob/f484436e744bcffd35be1725f2e7b2384cd47403/docker/omlbotlookup/data_access.R#L404
default != not present in database?

2. `get_parameter_default` is now called for parameters it does not work for. See the old example:
```http://localhost:8746/?algo=classif.ranger&task=3&parameters={%22num.trees%22:[1166],%22replace%22:[true],%22sample.fraction%22:[0.2773],%22mtry%22:[0.2345],%22respect.unordered.factors%22:[false],%22min.node.size%22:[0.0938]}```

**Comment by mboecker** [comment]
This is caused in Ranger by setup ids 220145 and 220233.
Both are missing the parameter sample.fraction.
It states [here](https://cran.r-project.org/web/packages/ranger/ranger.pdf) that the default should be

> sample.fraction
> Fraction of observations to sample. Default is 1 for sampling with replacement and 0.632 for sampling without replacement.

This default is simply not encoded in our server. Should I add it?

**Comment by mboecker** [comment]
We could also just throw these out.

**Comment by mboecker** [comment]
All runs between https://www.openml.org/r/2087629 and 2087638 are using these setups.

**Comment by mboecker** [comment]
This can't be right.
With the link in the topmost post here i get

> Removing 82.328364% (55174/67017) of runs, because they had missing values.

We agreed on removing every run with any NA and removing the insert-default-value-code, right?

**Comment by jakob-r** [comment]
Does this mean that 82% do not have values for `sample.fraction` or why are they beeing removed?

> removing the insert-default-value-code

That was the idea because we can not entirely be sure that our defaults match the defaults that actually have been used.

> removing every run with any NA 

Removing run where a parameter is not set although we expect it to be set (e.g. `sample.fraction`), yes.
Others -- i gues no?

**Comment by mboecker** [comment]
So I treated every missing row in the database as a NA now, removing replacing of NAs with defaults entirely.
That means, that 82% of runs hat any parameter missing, therefore had an NA, and therefore were removed by `complete.cases`.

I think we could add some default parameters back in, like those encoded in mlr directly.

**Comment by mboecker** [comment]
19999 runs of task 3 don't have min.node.size set. Thats 30% of the 67017 runs that we have.
This can't be due to reference runs, can it?

**Comment by PhilippPro** [comment]
This is because initially we did not include the min.node.size parameter but added it afterwards. ;)

**Comment by mboecker** [comment]
But what about respect.unordered.factors. There are only TRUEs in the database (50%) and the others are not set.

**Comment by PhilippPro** [comment]
Because FALSE is the default.

**Comment by mboecker** [comment]
OK so then we have to fill in the defaults for certain parameters, and drop the run if there is a NA for others.

**Comment by jakob-r** [comment]
Ok. So here it actually makes sense to put in the default

> Minimal node size. Default 1 for classification, 5 for regression, 3 for survival, and 10 for probability.

Pred type was probability or just lables?

**Comment by PhilippPro** [comment]
Probability always.

**Comment by jakob-r** [comment]
But then this line
https://github.com/ja-thomas/OMLbots/blob/5eb383cb7bc52c7ee292ea8a2a64e70f7a8dba12/snapshot_database/database_extraction.R#L94
would be wrong according to the documentation of ranger, right?

**Comment by PhilippPro** [comment]
Maybe you are right. But we exclude this runs afterwards anyway in 

[https://github.com/ja-thomas/OMLbots/blob/5eb383cb7bc52c7ee292ea8a2a64e70f7a8dba12/snapshot_database/database_extraction.R#L94](https://github.com/ja-thomas/OMLbots/blob/5eb383cb7bc52c7ee292ea8a2a64e70f7a8dba12/snapshot_database/database_extraction.R#L238)

It would be better to delete them in a different way. ;)

**Comment by jakob-r** [comment]
So after all you are throwing away all runs where min.node.size was not set explicitly?

**Comment by PhilippPro** [comment]
Yes. As we only have 500 000 results per algorithm, this is not a problem.

**Comment by jakob-r** [comment]
So @mboecker So let's do the same. Drop all runs of ranger where min.node.size is not set explicitly - preferably in `mysql_export(_task3).sql`.
If this is hard to do you probably can determine the first id after which the settings include min.node.size

**Comment by mboecker** [comment]
OK, and what about the other parameters? Should we fill in the defaults for NAs there?

**Comment by PhilippPro** [comment]
No, the reference runs were only RF runs as a "default performance" of "default "good" learner".

**Comment by mboecker** [comment]
But I said [here](https://github.com/mboecker/lookup-server/issues/36#issuecomment-393825960) that there are no `respect.unordered.factors = FALSE` runs in the database. The user could not query our server for runs with `respect.unordered.factors = FALSE`, if we fill in no defaults.

**Comment by jakob-r** [comment]
Sorry I oversaw one comment here from @mboecker 

> But what about respect.unordered.factors. There are only TRUEs in the database (50%) and the others are not set.

@PhilippPro responded with

> Because FALSE is the default.

BUT: After our initial confusion whether defaults get deleted I was assured that even if you set a parameter to it's default value it will be stored as such in the OML Database. In other words default value deletion should not happen. Now you telling us otherwise. Do you know why? Can we reproduce this?

**Comment by jakob-r** [comment]
Found the bug: https://github.com/openml/openml-r/commit/171aee917553e072a61de6d52fa0a7a1a4f6e715

All OpenMLBot runs seem to be affected. Luckily it only affects the "easy" defaults.
